src/d/actor/d_a_ship.cpp

```diff
@@ -5,460 +5,4713 @@
+    }
+    return;
 }
 
 /* 00002FBC-000031E4       .text setCrashData__8daShip_cFs */
-void daShip_c::setCrashData(short) {
-    /* Nonmatching */
+void daShip_c::setCrashData(short param1) {
+    float fVar1;
+    float fVar2;
+    daPy_lk_c *link;
+    s16 iVar5;
+    if (dComIfGp_event_runCheck() == 0) {
+        link = daPy_getPlayerLinkActorClass();
+        if (!link->checkNoDamageMode() && !link->checkNoControll()) {
+            if (!checkForceMove()) {
+                link->onShipDrop(param1);
+                procWait_init();
+            }
+        }
+        onStateFlg(daSFLG_UNK8000000_e);
+        if (m03B6 == 0) {
+            if (speedF > l_HIO.wind_inc_speed) {
+                speedF = l_HIO.wind_inc_speed;
+            }
+            speedF *= l_HIO.ef_speed_rate;
+            iVar5 = param1 - shape_angle.y;
+            fVar1 = cM_ssin(iVar5);
+            fVar2 = cM_scos(iVar5);
+            m0370 += fVar2 * 3072.0f;
+            m0374 += fVar2 * 500.0f;
+            m0372 -= fVar1 * 3072.0f;
+            m0376 -= fVar1 * 500.0f;
+            m03B6 = 30;
+        }
+        else {
+            m03B6--;
+        }
+    }
+    return;
 }
 
 /* 000031E4-00003490       .text checkNextMode__8daShip_cFi */
-void daShip_c::checkNextMode(int) {
-    /* Nonmatching */
+BOOL daShip_c::checkNextMode(int param_1) {
+    cXyz local_8;
+    cXyz local_20;
+    cXyz local_14;
+    
+    if (checkStateFlg(daSFLG_UNK4_e) != 0) {
+        setCrashData(shape_angle.y + 0x8000);
+        offStateFlg(daSFLG_UNK4_e);
+        return TRUE;
+    }
+
+    if (!dComIfGp_checkPlayerStatus0(0, daPyStts0_SHIP_RIDE_e)) {
+        if (m034D == 5) {
+            procGetOff_init();
+        }
+        else {
+            procWait_init();
+        }
+        return TRUE;
+    }
+    
+    if (dComIfGp_event_runCheck() == 0) {
+        cXyz* tgRVecP;
+        cXyz* tgHitPos;
+        dCcD_GObjInf* tgHitObj;
+        
+        tgRVecP = NULL;
+        tgHitPos = NULL;
+        tgHitObj = NULL;
+
+        if (mSph.ChkTgHit()) {
+            tgRVecP = mSph.GetTgRVecP();
+            tgHitPos = mSph.GetTgHitPosP();
+            tgHitObj = mSph.GetTgHitGObj();
+            
+        }
+        else {
+            dCcD_Cyl* cyl = mCyl;
+            for(int i = 0; i < 3; i++, cyl++) {
+                if (cyl->ChkTgHit()) {
+                    tgRVecP = cyl->GetTgRVecP();
+                    tgHitPos = cyl->GetTgHitPosP();
+                    tgHitObj = cyl->GetTgHitGObj();
+                }
+            }
+        }
+
+        if (tgHitObj && (tgHitObj->GetAtType() - AT_TYPE_WIND) == 0) {
+            if (dComIfGp_checkPlayerStatus0(0, daPyStts0_SHIP_RIDE_e)) {
+                onStateFlg((daSHIP_SFLG)(daSFLG_UNK1_e | daSFLG_UNK2000000_e));
+            }
+        }
+        else {
+            if (tgRVecP) {
+                short sVar2;
+                local_14.set(tgRVecP->x, 0.0f, tgRVecP->z);
+                if (local_14.abs2() < 0.1f) {
```

Comment:
I see several places where abs() or abs2() are called that have the x, 0, z pattern before them so the XZ versions of the inlines should work here I think:
```suggestion
                if (tgRVecP->abs2XZ() < 0.1f) {
```
btw this might help improve the stack offsets you mentioned before